backup: TestShowBackupPathIsCollectionRoot failed

backup.TestShowBackupPathIsCollectionRoot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20426243?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20426243?buildTab=artifacts#/) on master @ [216504747d699dd6c0c4425540b653199d9048d3](https://github.com/cockroachdb/cockroach/commits/216504747d699dd6c0c4425540b653199d9048d3):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot3456414241
    test_log_scope.go:82: use -show-logs to present logs inline
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot3456414241
--- FAIL: TestShowBackupPathIsCollectionRoot (1.02s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54114

backup.TestShowBackupPathIsCollectionRoot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20426992?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20426992?buildTab=artifacts#/) on master @ [7b65b179304f8667c4cbd8d949208ba273195fb9](https://github.com/cockroachdb/cockroach/commits/7b65b179304f8667c4cbd8d949208ba273195fb9):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot3593990188
    test_log_scope.go:82: use -show-logs to present logs inline
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot3593990188
--- FAIL: TestShowBackupPathIsCollectionRoot (1.07s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestShowBackupPathIsCollectionRoot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20428750?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20428750?buildTab=artifacts#/) on master @ [eb8c0d98da22725422f12eac0f757772de9f0e00](https://github.com/cockroachdb/cockroach/commits/eb8c0d98da22725422f12eac0f757772de9f0e00):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot3059628631
    test_log_scope.go:82: use -show-logs to present logs inline
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot3059628631
--- FAIL: TestShowBackupPathIsCollectionRoot (0.83s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestShowBackupPathIsCollectionRoot [failed](https://mesolite.cluster.engflow.com/invocations/default/9a9a218f-c49b-40ea-a6e5-a19509c4fd75?testReportRun=3&testReportShard=29&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on master @ [eb8c0d98da22725422f12eac0f757772de9f0e00](https://github.com/cockroachdb/cockroach/commits/eb8c0d98da22725422f12eac0f757772de9f0e00):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestShowBackupPathIsCollectionRoot1560334713
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:156: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestShowBackupPathIsCollectionRoot1560334713
--- FAIL: TestShowBackupPathIsCollectionRoot (5.49s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=29</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestShowBackupPathIsCollectionRoot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20429055?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20429055?buildTab=artifacts#/) on master @ [65656d410f6f3158421b3eb4cf30eda00786398c](https://github.com/cockroachdb/cockroach/commits/65656d410f6f3158421b3eb4cf30eda00786398c):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot1150946305
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:156: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot1150946305
--- FAIL: TestShowBackupPathIsCollectionRoot (2.49s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestShowBackupPathIsCollectionRoot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20427240?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20427240?buildTab=artifacts#/) on master @ [eb8c0d98da22725422f12eac0f757772de9f0e00](https://github.com/cockroachdb/cockroach/commits/eb8c0d98da22725422f12eac0f757772de9f0e00):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/ac96935e425162500fbbaaa5d00c5635/logTestShowBackupPathIsCollectionRoot924426876
    test_log_scope.go:82: use -show-logs to present logs inline
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ac96935e425162500fbbaaa5d00c5635/logTestShowBackupPathIsCollectionRoot924426876
--- FAIL: TestShowBackupPathIsCollectionRoot (1.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestShowBackupPathIsCollectionRoot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20429220?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20429220?buildTab=artifacts#/) on master @ [6766271f0df9b5f5150ee3bd87df95465dcb0817](https://github.com/cockroachdb/cockroach/commits/6766271f0df9b5f5150ee3bd87df95465dcb0817):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot1828677276
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:156: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f64d67a414749afd3c16ee3b3d9bc76/logTestShowBackupPathIsCollectionRoot1828677276
--- FAIL: TestShowBackupPathIsCollectionRoot (1.91s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestShowBackupPathIsCollectionRoot [failed](https://mesolite.cluster.engflow.com/invocations/default/54d2dd09-c138-4b14-898f-c1a02ebf579b?testReportRun=2&testReportShard=29&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on master @ [eb8c0d98da22725422f12eac0f757772de9f0e00](https://github.com/cockroachdb/cockroach/commits/eb8c0d98da22725422f12eac0f757772de9f0e00):


```
=== RUN   TestShowBackupPathIsCollectionRoot
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestShowBackupPathIsCollectionRoot2696624506
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:156: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    show_test.go:719: expected query 'SHOW BACKUP '' IN $1' error 'The specified path is the root of a backup collection.', got: pq: parsing subdir "" for flattening: parsing time "" as "/2006/01/02-150405.00": cannot parse "" as "/"
    testutils.go:307: no Invalid Descriptors
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestShowBackupPathIsCollectionRoot2696624506
--- FAIL: TestShowBackupPathIsCollectionRoot (7.73s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=29</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestShowBackupPathIsCollectionRoot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/backup/backupdest/backup_index.go
+++ b/pkg/backup/backupdest/backup_index.go
@@ -440,7 +440,9 @@ func indexSubdir(subdir string) (string, error) {
 func flattenSubdirForIndex(subdir string) (string, error) {
 	subdirTime, err := time.Parse(backupbase.DateBasedIntoFolderName, subdir)
 	if err != nil {
-		return "", errors.Wrapf(err, "parsing subdir %q for flattening", subdir)
+		return "", errors.Wrapf(
+			err, "subdir does not match format '%s'", backupbase.DateBasedIntoFolderName,
+		)
 	}
 	return subdirTime.Format(backupbase.BackupIndexFlattenedSubdir), nil
 }
@@ -450,7 +452,9 @@ func flattenSubdirForIndex(subdir string) (string, error) {
 func unflattenIndexSubdir(flattened string) (string, error) {
 	subdirTime, err := time.Parse(backupbase.BackupIndexFlattenedSubdir, flattened)
 	if err != nil {
-		return "", errors.Wrapf(err, "parsing flattened index subdir %q for unflattening", flattened)
+		return "", errors.Wrapf(
+			err, "index subdir does not match format %s", backupbase.BackupIndexFlattenedSubdir,
+		)
 	}
 	unflattened := subdirTime.Format(backupbase.DateBasedIntoFolderName)
 	return unflattened, nil

--- a/pkg/backup/show_test.go
+++ b/pkg/backup/show_test.go
@@ -713,12 +713,15 @@ func TestShowBackupPathIsCollectionRoot(t *testing.T) {
 	_, sqlDB, _, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, InitManualReplication)
 	defer cleanupFn()
 
+	// Error output changes depending on whether or not the index is used. This
+	// deterministically enables the index to make the error output consistent.
+	sqlDB.Exec(t, "SET CLUSTER SETTING backup.index.read.enabled = true")
+
 	// Make an initial backup.
 	sqlDB.Exec(t, `BACKUP data.bank INTO $1`, localFoo)
 
 	// Ensure proper error gets returned from back SHOW BACKUP Path
-	sqlDB.ExpectErr(t, "The specified path is the root of a backup collection.",
-		"SHOW BACKUP '' IN $1", localFoo)
+	sqlDB.ExpectErr(t, "subdir does not match format", "SHOW BACKUP '' IN $1", localFoo)
 }
 
 // TestShowBackupCheckFiles verifies the check_files option catches a corrupt

